### PR TITLE
Document code review process at OKI

### PR DESCRIPTION
**Pull request by pwalsh**

### CODE REVIEW

File: sections/reviews.md [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+## How
+
+The process for code reviews is straight forward as long as you are already following our coding standards for working with version control, and in particular our git flow-influenced process of branch management. The bottom line is that, before any code hits `master`, it must be reviewed by a colleague.
+
+### Submitting code for review
+
+Submitting code for review should follow these steps:
+
+1. Work out of your own fork of the main repository (usually `origin`).
```
Comment (akariv) [inline sections/reviews.md]: Is that really necessary? You can submit PRs from the feature branch to the master branch even if the are in the same repo.

Comment (pwalsh) [inline sections/reviews.md]: I usually do exactly as you describe. I'm easy.

File: sections/reviews.md [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+2. Always, always work from a feature or fix branch that is checked out of `master`.
+3. Each feature or fix branch should be focused on a discrete unit of work.
+4. When your unit of work is complete, submit a pull request against the `master` branch on `origin`.
```
Comment (akariv) [inline sections/reviews.md]: Make sure the PR has a description explaining the nature of the change, in technical terms.

i.e. not 'fixing the multi-view feature' but 'adding exception handling to the DB migrations'

(no, the name of the branch or title of the commit are usually not enough...)

File: sections/reviews.md [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+5. Wait for the CI server to run, validating your tests pass on all target environments. If you do not have tests, or have not configured CI, do that immediately before proceeding.
+6. If CI is green, ask a colleague to review your pull request (we will automate this in the near future). This preferably will be a colleague who is not working on your project.
```
Comment (akariv) [inline sections/reviews.md]: Suggestion: Let's have a CR channel in slack which you could post PR urls in there, and whoever's available can take the review (better than all the coordination described here IMO).

Comment (akariv) [inline sections/reviews.md]: (we could have a slackbot to do that, naturally, reducing a little more boilerplate)

Comment (pwalsh) [inline sections/reviews.md]: We will do that slackbot/automation.

File: sections/reviews.md (hunk 1 above)
Comment (brew) [inline sections/reviews.md]: I think either is okay, as long as old feature branches are removed from the origin repo when no longer needed.

Comment (pwalsh) [inline sections/reviews.md]: @brew yes that is the main annoying thing. In the last year I removed something like 20 branches from both OpenSpending and Open Data Census because people just merged PRs and left their mess around :)

Comment (brew) [inline sections/reviews.md]: Perhaps we should add this to the coding-standards: It's the merger's responsibility to remove redundant feature branches. 

btw - git cli can be useful for periodic branch pruning with `git branch --merged` and `git branch --no-merged`.

File: sections/reviews.md [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+7. Address any questions from the code review. Sometimes this will involving refactoring, other times it will just mean answering questions.
+8. When the reviewer indicates that the pull request is ready to merge, you may merge into `master`.
```
Comment (amercader) [inline sections/reviews.md]: So you merge your own PR? Or is the reviewer who does it?

Comment (pwalsh) [inline sections/reviews.md]: As long as the reviewer indicates that the pull request is ready, I don't think it matters who "clicks the button". The reason that I do not expect the reviewer to actually merge is because the reviewer is not necessarily working on the codebase and so perhaps in these circumstances it is better to let the developer chose the time for merge.

We can change it to ask reviewer to always merge, on the assumption that the developer is expecting this to happen when the code is approved.

File: sections/reviews.md [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+* Is the code idiomatic for the language and/or framework it is written in?
+
+There are some clear red flags that, on seeing any of them, you should immediately ask your colleague to address before continuing with any code review:
+
+* There is no CI
+* There are no tests
+* There are tests that fail
+* There is code that is not linted
+
+### Prioritizing code reviews
```
Comment (amercader) [inline sections/reviews.md:47]: Frankly, I'm not convinced at all by this. I appreciate the intention behind it and fully acknowledge the importance of the reviews but the reality is that we have packed schedules and are already distracted by other "small little things" during the day. I don't know how feasible is to give top priority to this.
Maybe I missed the conversation about this that you mention, but is there a sense on how much should you spend on reviews (per day/ week, on a particular one?)

Comment (pwalsh) [inline sections/reviews.md:47]: We are all busy, all the time :). But I hear you on the small distractions. Let's see how it goes.

Comment (vitorbaptista) [inline sections/reviews.md:47]: I understand the reasoning on giving top priority for this. There were a few times where I got one pull request that depended on another pull request that depended on another one. When there're changes needed in the first PR in the chain, it gets messy very quickly.

Still, maybe reviewing once per day would be a good compromise? I'd be happy to try either way to see how it goes.

Comment (akariv) [inline sections/reviews.md:47]: I think that if we change this to a pull model, the distraction problem will be more-or-less solved.
That is - whenever you're already out of context (e.g. right after a conference call or lunch), take a look at the #CodeReview channel in slack and see if there's any CR pending. You can just do one or two and then dive back into your code.

File: sections/reviews.md (hunk 1 above)
Comment (pwalsh) [inline sections/reviews.md]: @brew thanks for the branch pruning tip!

File: sections/reviews.md (hunk 5 above)
Comment (pwalsh) [inline sections/reviews.md:47]: @akariv understood, but I don't want to move to a pull model now. Perhaps once we have integrated code reviews into our daily work practice fully.

File: sections/reviews.md [hunk 6]
```diff
@@ -0,0 +1,74 @@
+# Code reviews
+
+At Open Knowledge International (OKI) all code must be reviewed by a colleague before merging into the main branch of the repository (usually `master`). This short document examples how to do code reviews, and why we do it.
+
+## How
+
+The process for code reviews is straight forward as long as you are already following our coding standards for working with version control, and in particular our git flow-influenced process of branch management. The bottom line is that, before any code hits `master`, it must be reviewed by a colleague.
```
Comment (brew) [inline sections/reviews.md:7]: _must_ is very strong. I still have concerns that this stricture doesn't fit in with all projects. Especially in the early stages of a new project.
